#Question 1: Loan Eligibility Calculator (Functions)

**Description**

Create a loan eligibility calculator where different functions handle income, credit score, and age to
determine eligibility.

**Requirements**

Create separate functions for:
- check_income(income): Validates if income is above minimum threshold (Rs
25,000/month)
- check_credit_score(score): Validates if credit score is above 650
- check_age(age): Validates if age is between 21 and 65
- calculate_loan_amount(income, credit_score): Calculates eligible loan amount
based on:
• Income × 60 (if credit score  750)
• Income × 50 (if credit score 700-749)
• Income × 40 (if credit score 650-699)
• determine_eligibility(income, credit_score, age): Main function that uses all
above functions to determine final eligibility
Display detailed eligibility report

**Input/Output Example**

Enter monthly income: 50000

Enter credit score: 720

Enter age: 30

=== Loan Eligibility Report ===

Income Check: ✓ Passed (Rs 50,000)

Credit Score Check: ✓ Passed (720)

Age Check: ✓ Passed (30 years)

Eligible Loan Amount: Rs 25,00,000

Status: ELIGIBLE

**Constraints**

- All functions must be separate and reusable
- Handle invalid inputs (negative values, out of range)
- Return appropriate boolean or numeric values


In [1]:
def check_income(income):
    return income >= 25000

def check_credit_score(score):
    return score >= 650

def check_age(age):
    return 21 <= age <= 65

def calculate_loan_amount(income, credit_score):
    if credit_score >= 750:
        return income * 60

    elif credit_score >= 700:
        return income * 50

    elif credit_score >= 650:
        return income * 40

    else:
        return 0

def determine_eligibility(income, credit_score, age):

    income_status = check_income(income)
    credit_status = check_credit_score(credit_score)
    age_status = check_age(age)

    if income_status and credit_status and age_status:
        loan_amount = calculate_loan_amount(
            income,
            credit_score
        )
        return True, loan_amount
    return False, 0

try:

    income = float(input("Enter monthly income: "))
    credit_score = int(input("Enter credit score: "))
    age = int(input("Enter age: "))
    if income < 0 or credit_score < 0 or age < 0:
        print("Error: Values cannot be negative.")

    else:

        eligible, loan_amount = determine_eligibility(
            income,
            credit_score,
            age
        )
        print("\n=== Loan Eligibility Report ===")
        if check_income(income):
            print(f"Income Check: ✓ Passed (Rs {income:,.0f})")
        else:
            print(f"Income Check: ✗ Failed (Rs {income:,.0f})")

        if check_credit_score(credit_score):
            print(f"Credit Score Check: ✓ Passed ({credit_score})")
        else:
            print(f"Credit Score Check: ✗ Failed ({credit_score})")

        if check_age(age):
            print(f"Age Check: ✓ Passed ({age} years)")
        else:
            print(f"Age Check: ✗ Failed ({age} years)")

        if eligible:
            print(f"\nEligible Loan Amount: Rs {loan_amount:,.0f}")
            print("Status: ELIGIBLE")
        else:
            print("\nEligible Loan Amount: Rs 0")
            print("Status: NOT ELIGIBLE")

except ValueError:
    print("Error: Please enter valid numeric values.")

Enter monthly income: 50000
Enter credit score: 720
Enter age: 30

=== Loan Eligibility Report ===
Income Check: ✓ Passed (Rs 50,000)
Credit Score Check: ✓ Passed (720)
Age Check: ✓ Passed (30 years)

Eligible Loan Amount: Rs 2,500,000
Status: ELIGIBLE


#Question 2: Function Parameters - Flexible Calculator

**Description**
Create a calculator that demonstrates different types of function parameters:

default arguments,

args, and *kwargs.

**Requirements**

Create a function calculate() that:
- Accepts a required operation parameter ('add', 'subtract', 'multiply', 'divide')
- Uses args to accept variable number of numbers
- Uses *kwargs for optional parameters:
• round_to: Number of decimal places (default: 2)
• display: Whether to display result (default: True)

**Input/Output Example**

Example 1: Using *args

calculate('add', 10, 20, 30, 40)

Output: Sum: 100.00

Example 2: Using default and **kwargs

calculate('multiply', 5, 6, 7, round_to=0, display=True)

Output: Product: 210

**Constraints**

- Handle division by zero
- Validate operation types
- Use appropriate parameter order (required, args, *kwargs)

In [2]:
def calculate(operation, *args, **kwargs):
    round_to = kwargs.get("round_to", 2)
    display = kwargs.get("display", True)

    if len(args) == 0:
        print("Error: No numbers provided.")
        return

    if operation == "add":
        result = sum(args)
        label = "Sum"

    elif operation == "subtract":
        result = args[0]
        for num in args[1:]:
            result -= num
        label = "Difference"

    elif operation == "multiply":
        result = 1
        for num in args:
            result *= num
        label = "Product"

    elif operation == "divide":
        result = args[0]
        for num in args[1:]:
            if num == 0:
                print("Error: Division by zero.")
                return
            result /= num
        label = "Quotient"

    else:
        print("Invalid operation.")
        return

    result = round(result, round_to)

    if display:
        print(f"{label}: {result}")

    return result

In [3]:
calculate("add", 10, 20, 30, 40)

Sum: 100


100

In [4]:
calculate('multiply', 5, 6, 7, round_to=0, display=True)

Product: 210


210

#Question 3: Nested Functions - Invoice Calculator


**Description**

Create an invoice calculation system using nested functions.

**Requirements**

Create a main function calculate_invoice() that:
- Accepts base price, tax rate, and optional discount code
- Contains nested functions:
• calculate_tax(price, rate): Calculates tax amount
• apply_discount(price, code): Applies discount based on code
• "DISCOUNT10": 10% off
• "DISCOUNT20": 20% off
• "VIP": 25% off
• format_currency(amount): Formats amount with currency symbol
• Returns formatted invoice with breakdown

Display itemized invoice

**Input/Output Example**

calculate_invoice(1000, 5, "DISCOUNT10")

=== Invoice ===

Base Price: Rs 1,000.00

Discount (10%): -Rs 100.00

Subtotal: Rs 900.00

Tax (5%): Rs 45.00

Total: Rs 945.00

**Constraints**

- Nested functions should only be accessible within main function
- Validate discount codes
- Handle invalid inputs

In [5]:
def calculate_invoice(base_price, tax_rate, discount_code=None):
    if base_price < 0:
        return "Error: Base price cannot be negative."

    if tax_rate < 0:
        return "Error: Tax rate cannot be negative."

    def calculate_tax(price, rate):
        return price * rate / 100

    def apply_discount(price, code):

        discounts = {
            "DISCOUNT10": 10,
            "DISCOUNT20": 20,
            "VIP": 25
        }
        if code is None:
            return price, 0

        if code not in discounts:
            return price, 0

        discount_percent = discounts[code]

        discount_amount = price * discount_percent / 100

        return price - discount_amount, discount_percent

    def format_currency(amount):
        return f"Rs {amount:,.2f}"

    subtotal, discount_percent = apply_discount(
        base_price,
        discount_code
    )
    discount_amount = base_price - subtotal

    tax_amount = calculate_tax(
        subtotal,
        tax_rate
    )

    total = subtotal + tax_amount

    print("=== Invoice ===")
    print(f"Base Price: {format_currency(base_price)}")

    if discount_percent > 0:
        print(
            f"Discount ({discount_percent}%): "
            f"-{format_currency(discount_amount)}"
        )
    else:
        print("Discount: Rs 0.00")

    print(f"Subtotal: {format_currency(subtotal)}")
    print(f"Tax ({tax_rate}%): {format_currency(tax_amount)}")
    print(f"Total: {format_currency(total)}")

In [6]:
calculate_invoice(1000, 5, "DISCOUNT10")

=== Invoice ===
Base Price: Rs 1,000.00
Discount (10%): -Rs 100.00
Subtotal: Rs 900.00
Tax (5%): Rs 45.00
Total: Rs 945.00


#Question 4: Functions Returning Functions - Currency Converter Factory


**Description**

Create a currency converter system where a function returns another function.

**Requirements**

Create a function currency_converter_factory(rate) that:
- Accepts exchange rate
- Returns a converter function that converts amounts using that rate

Create multiple converters:
- USD to INR (rate: 83.0)
- EUR to INR (rate: 90.0)
- GBP to INR (rate: 105.0)

Create a function create_converter_menu() that returns a dictionary of converter functions

Allow users to convert amounts using selected converter

**Input/Output Example**

USD to INR Converter created (Rate: 83.0)

EUR to INR Converter created (Rate: 90.0)

GBP to INR Converter created (Rate: 105.0)

Converting $100 USD to INR: Rs 8,300.00

Converting €50 EUR to INR: Rs 4,500.00

Converting £30 GBP to INR: Rs 3,150.00

**Constraints**

- Validate exchange rates (must be positive)
- Handle negative amounts appropriately

In [7]:
def currency_converter_factory(rate):
    if rate <= 0:
        raise ValueError(
            "Exchange rate must be positive."
        )
    def converter(amount):
        if amount < 0:
            return "Error: Amount cannot be negative."
        return amount * rate
    return converter
def create_converter_menu():
    return {
        "USD": currency_converter_factory(83.0),
        "EUR": currency_converter_factory(90.0),
        "GBP": currency_converter_factory(105.0)
    }
converters = create_converter_menu()
print("USD to INR Converter created (Rate: 83.0)")
print("EUR to INR Converter created (Rate: 90.0)")
print("GBP to INR Converter created (Rate: 105.0)")
usd_amount = 100
eur_amount = 50
gbp_amount = 30

usd_inr = converters["USD"](usd_amount)
eur_inr = converters["EUR"](eur_amount)
gbp_inr = converters["GBP"](gbp_amount)

print(
    f"\nConverting ${usd_amount} USD to INR: "
    f"Rs {usd_inr:,.2f}"
)

print(
    f"Converting €{eur_amount} EUR to INR: "
    f"Rs {eur_inr:,.2f}"
)

print(
    f"Converting £{gbp_amount} GBP to INR: "
    f"Rs {gbp_inr:,.2f}"
)

USD to INR Converter created (Rate: 83.0)
EUR to INR Converter created (Rate: 90.0)
GBP to INR Converter created (Rate: 105.0)

Converting $100 USD to INR: Rs 8,300.00
Converting €50 EUR to INR: Rs 4,500.00
Converting £30 GBP to INR: Rs 3,150.00


#Question 5: Lambda Functions - Data Processing

**Description**

Use lambda functions with map(), filter(), and reduce() to process data.

**Requirements**

Using map():
- Square all numbers in a list
- Convert temperatures from Celsius to Fahrenheit
- Extract first letter from list of names

Using filter():
- Filter even numbers
- Filter names starting with specific letter
- Filter products above certain price

Using reduce():
- Calculate sum of numbers
- Find maximum number
- Calculate product of numbers

Custom Sorting:
- Sort list of dictionaries by specific key using lambda
- Sort students by grade (descending)

**Input/Output Example**

Numbers: [1, 2, 3, 4, 5]

Squared: [1, 4, 9, 16, 25]

Temperatures (°C): [0, 25, 37, 100]

Temperatures (°F): [32.0, 77.0, 98.6, 212.0]

Numbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Even numbers: [2, 4, 6, 8, 10]

Students sorted by grade:
[{'name': 'Alice', 'grade': 95}, {'name': 'Bob', 'grade': 87}, {'name': 'Charlie', 'grade': 78

**Constraints**
- Import functools for reduce()
- Use lambda functions, not regular functions
- Handle empty lists appropriately

In [8]:
from functools import reduce

print("=== Lambda Functions with map(), filter(), reduce() ===\n")

numbers = [1, 2, 3, 4, 5]

squared = list(map(lambda x: x**2, numbers))

print("Numbers:", numbers)
print("Squared:", squared)

temperatures_c = [0, 25, 37, 100]

temperatures_f = list(
    map(lambda c: (c * 9/5) + 32, temperatures_c)
)

print("\nTemperatures (°C):", temperatures_c)
print("Temperatures (°F):", temperatures_f)

names = ["Alice", "Bob", "Charlie", "David"]

first_letters = list(
    map(lambda name: name[0], names)
)

print("\nNames:", names)
print("First Letters:", first_letters)
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

even_numbers = list(
    filter(lambda x: x % 2 == 0, numbers)
)

print("\nNumbers:", numbers)
print("Even Numbers:", even_numbers)

names = ["Alice", "Andrew", "Bob", "Charlie", "Alex"]

a_names = list(
    filter(lambda name: name.startswith("A"), names)
)

print("\nNames:", names)
print("Names Starting with A:", a_names)

products = [
    {"name": "Laptop", "price": 50000},
    {"name": "Mouse", "price": 500},
    {"name": "Keyboard", "price": 1500},
    {"name": "Monitor", "price": 12000}
]

expensive_products = list(
    filter(lambda product: product["price"] > 1000, products)
)

print("\nProducts Above Rs 1000:")
print(expensive_products)

numbers = [1, 2, 3, 4, 5]

total = reduce(
    lambda x, y: x + y,
    numbers
)

maximum = reduce(
    lambda x, y: x if x > y else y,
    numbers
)

product = reduce(
    lambda x, y: x * y,
    numbers
)

print("\nNumbers:", numbers)
print("Sum:", total)
print("Maximum:", maximum)
print("Product:", product)

students = [
    {"name": "Alice", "grade": 95},
    {"name": "Bob", "grade": 87},
    {"name": "Charlie", "grade": 78},
    {"name": "David", "grade": 91}
]

students_sorted = sorted(
    students,
    key=lambda student: student["grade"],
    reverse=True
)

print("\nStudents Sorted by Grade (Descending):")
print(students_sorted)

=== Lambda Functions with map(), filter(), reduce() ===

Numbers: [1, 2, 3, 4, 5]
Squared: [1, 4, 9, 16, 25]

Temperatures (°C): [0, 25, 37, 100]
Temperatures (°F): [32.0, 77.0, 98.6, 212.0]

Names: ['Alice', 'Bob', 'Charlie', 'David']
First Letters: ['A', 'B', 'C', 'D']

Numbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Even Numbers: [2, 4, 6, 8, 10]

Names: ['Alice', 'Andrew', 'Bob', 'Charlie', 'Alex']
Names Starting with A: ['Alice', 'Andrew', 'Alex']

Products Above Rs 1000:
[{'name': 'Laptop', 'price': 50000}, {'name': 'Keyboard', 'price': 1500}, {'name': 'Monitor', 'price': 12000}]

Numbers: [1, 2, 3, 4, 5]
Sum: 15
Maximum: 5
Product: 120

Students Sorted by Grade (Descending):
[{'name': 'Alice', 'grade': 95}, {'name': 'David', 'grade': 91}, {'name': 'Bob', 'grade': 87}, {'name': 'Charlie', 'grade': 78}]


#Question 6: Decorators - Function Enhancement

**Description**

Create and use decorators to add functionality to functions.

**Requirements**

Timing Decorator:
- Create @timing_decorator that measures function execution time
- Display time taken in seconds

Logging Decorator:
- Create @log_decorator that logs function calls
- Log function name, arguments, and return value

Validation Decorator:
- Create @validate_input that validates function arguments
- Check if numbers are positive
- Check if strings are non-empty

Retry Decorator:
- Create @retry_decorator(max_attempts) that retries function on failure
- Retry up to max_attempts times

Apply multiple decorators to a single function

**Input/Output Example**

@timing_decorator

@log_decorator

def calculate_factorial(n): #implementation

pass

calculate_factorial(5)

Output:

[LOG] Calling function: calculate_factorial

[LOG] Arguments: (5,)

[LOG] Return value: 120

[LOG] Function calculate_factorial took 0.0001 seconds to execute.

**Constraints**

- Use functools.wraps to preserve function metadata
- Handle decorator arguments properly
- Test with various function signature

In [9]:
import time
from functools import wraps

def timing_decorator(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        start_time = time.time()
        result = func(*args, **kwargs)
        end_time = time.time()
        print(
            f"[LOG] Function {func.__name__} "
            f"took {end_time - start_time:.6f} seconds to execute."
        )
        return result
    return wrapper

def log_decorator(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        print(f"[LOG] Calling function: {func.__name__}")
        print(f"[LOG] Arguments: {args}")
        result = func(*args, **kwargs)
        print(f"[LOG] Return value: {result}")
        return result
    return wrapper

def validate_input(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        for arg in args:
            if isinstance(arg, (int, float)):
                if arg <= 0:
                    raise ValueError(
                        "Numeric arguments must be positive."
                    )
            if isinstance(arg, str):
                if arg.strip() == "":
                    raise ValueError(
                        "String arguments cannot be empty."
                    )
        return func(*args, **kwargs)
    return wrapper

def retry_decorator(max_attempts):
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            attempts = 0
            while attempts < max_attempts:

                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    attempts += 1
                    print(
                        f"[RETRY] Attempt {attempts} failed: {e}"
                    )
                    if attempts == max_attempts:
                        raise
        return wrapper
    return decorator

@timing_decorator
@log_decorator
@validate_input
def calculate_factorial(n):

    result = 1

    for i in range(1, n + 1):
        result *= i

    return result

counter = 0

@retry_decorator(3)
def unstable_function():

    global counter

    counter += 1

    if counter < 3:
        raise Exception("Temporary failure")

    return "Success"

print("\n=== Factorial Example ===")

print("Factorial:", calculate_factorial(5))

print("\n=== Retry Example ===")

print(unstable_function())


=== Factorial Example ===
[LOG] Calling function: calculate_factorial
[LOG] Arguments: (5,)
[LOG] Return value: 120
[LOG] Function calculate_factorial took 0.000036 seconds to execute.
Factorial: 120

=== Retry Example ===
[RETRY] Attempt 1 failed: Temporary failure
[RETRY] Attempt 2 failed: Temporary failure
Success


#Question 7: Custom Module Creation

**Description**

Create a custom module with utility functions and use it in a main program.

**Requirements**

Create a module math_utils.py with:
- calculate_percentage(part, whole): Calculate percentage
- calculate_average(numbers): Calculate average
- find_max_min(numbers): Find maximum and minimum
- is_prime(n): Check if number is prime

Create a module string_utils.py with:
- reverse_string(text): Reverse a string
- count_words(text): Count words in text
- capitalize_words(text): Capitalize each word
- remove_duplicates(text): Remove duplicate characters

Create main.py that:
- Imports functions from both modules
- Uses different import methods (import, from...import, alias)
- Demonstrates usage of all functions

File Structure

project/

■■■ math_utils.py

■■■ string_utils.py

■■■ main.py

**Input/Output Example**

=== Math Utilities ===

Percentage: 75.0%

Average: 85.5

Max: 100, Min: 70

Is 17 prime? True

=== String Utilities ===

Reversed: !dlroW olleH

Word Count: 3

Capitalized: Hello World Python

No Duplicates: Helo Wrd

**Constraints**

- Add docstrings to all functions
- Handle edge cases (empty lists, None values)
- Use proper module structure

In [13]:
%%writefile math_utils.py

"""
math_utils.py
"""

def calculate_percentage(part, whole):
    if whole == 0:
        return 0
    return (part / whole) * 100

def calculate_average(numbers):
    if not numbers:
        return 0
    return sum(numbers) / len(numbers)

def find_max_min(numbers):
    if not numbers:
        return None, None
    return max(numbers), min(numbers)

def is_prime(n):
    if n <= 1:
        return False

    for i in range(2, int(n**0.5)+1):
        if n % i == 0:
            return False

    return True

Writing math_utils.py


In [14]:
%%writefile string_utils.py

"""
string_utils.py
"""

def reverse_string(text):
    return text[::-1]

def count_words(text):
    return len(text.split())

def capitalize_words(text):
    return text.title()

def remove_duplicates(text):
    seen = set()
    result = ""

    for ch in text:
        if ch not in seen:
            result += ch
            seen.add(ch)

    return result

Writing string_utils.py


In [16]:
# main.py

import math_utils
from string_utils import reverse_string
import string_utils as su

print("=== Math Utilities ===")

percentage = math_utils.calculate_percentage(
    75,
    100
)

print(f"Percentage: {percentage}%")

numbers = [70, 80, 92, 100]

average = math_utils.calculate_average(numbers)

print(f"Average: {average}")

maximum, minimum = math_utils.find_max_min(numbers)

print(f"Max: {maximum}, Min: {minimum}")

print(
    f"Is 17 prime? "
    f"{math_utils.is_prime(17)}"
)

print("\n=== String Utilities ===")

text = "Hello World!"

print(
    f"Reversed: "
    f"{reverse_string(text)}"
)

sentence = "hello world python"

print(
    f"Word Count: "
    f"{su.count_words(sentence)}"
)

print(
    f"Capitalized: "
    f"{su.capitalize_words(sentence)}"
)

print(
    f"No Duplicates: "
    f"{su.remove_duplicates(text)}"
)

=== Math Utilities ===
Percentage: 75.0%
Average: 85.5
Max: 100, Min: 70
Is 17 prime? True

=== String Utilities ===
Reversed: !dlroW olleH
Word Count: 3
Capitalized: Hello World Python
No Duplicates: Helo Wrd!


#Question 8: Package Creation and Organization

**Description**

Create a package structure with multiple modules and demonstrate package usage.

**Requirements**

Create a package calculator with the following structure:

calculator/ ■■■ init .py ■■■ basic.py (basic operations) ■■■
advanced.py (advanced operations) ■■■ scientific.py (scientific
operations)

init .py should:
- Import commonly used functions
- Define package version
- Export main functions

Each module should contain related functions:
- basic.py: add, subtract, multiply, divide
- advanced.py: power, sqrt, factorial
- scientific.py: sin, cos, log

Create main.py that uses the package

**Input/Output Example**

From init .py imports

from calculator import add, multiply, power

result1 = add(10, 20) # 30

result2 = power(2, 8) # 256

 Direct module imports

from calculator.advanced import factorial

result3 = factorial(5) # 120

 Using package

import calculator.scientific as sci

result4 = sci.sin(3.14159/2) # ~1.0

**Constraints**

- Proper init .py implementation
- Clear module organization
- Document package structure

In [21]:
!mkdir -p calculator

In [22]:
%%writefile calculator/basic.py

def add(a, b):
    return a + b

def subtract(a, b):
    return a - b

def multiply(a, b):
    return a * b

def divide(a, b):
    if b == 0:
        raise ValueError("Division by zero")
    return a / b

Writing calculator/basic.py


In [23]:
%%writefile calculator/advanced.py

import math

def power(a, b):
    return a ** b

def sqrt(x):
    return math.sqrt(x)

def factorial(n):
    return math.factorial(n)

Writing calculator/advanced.py


In [24]:
%%writefile calculator/scientific.py

import math

def sin(x):
    return math.sin(x)

def cos(x):
    return math.cos(x)

def log(x):
    return math.log(x)

Writing calculator/scientific.py


In [25]:
%%writefile calculator/__init__.py

__version__ = "1.0.0"

from .basic import add, subtract, multiply, divide
from .advanced import power, sqrt, factorial

__all__ = [
    "add",
    "subtract",
    "multiply",
    "divide",
    "power",
    "sqrt",
    "factorial"
]

Writing calculator/__init__.py


In [26]:
!find calculator

calculator
calculator/__init__.py
calculator/advanced.py
calculator/basic.py
calculator/scientific.py


In [27]:
from calculator import add, multiply, power

print(add(10, 20))
print(multiply(5, 6))
print(power(2, 8))

30
30
256


In [28]:
from calculator.advanced import factorial

print(factorial(5))

120


In [29]:
import calculator.scientific as sci

print(sci.sin(3.14159/2))

0.9999999999991198
